pii/ner_model_training/utils/chunking.py

```diff
@@ -0,0 +1,187 @@
+    """
+    return Dataset.from_list(
+        list(
+            itertools.chain(
+                *(chunk_inputs(**entry, tokenizer=tokenizer, max_length=tokenizer.max_len_single_sentence,
+                               overlap_freq=overlap_freq) for entry in tqdm(list(dataset)))
+            )
+        )
+    )
+
+
+def chunk_inputs(input_ids,
+                 attention_mask,
+                 id,
+                 tokenizer,
+                 max_length,
+                 labels=None,
+                 overlap_freq=0,
+                 **kwargs):
+    """
+    Processes a sequence of `input_ids`, `attention_mask`, and `labels` by dividing it
+    into smaller chunks to be processed by the model. The sequence is split into chunks
+    of a specified maximum length, with a defined overlap frequency `overlap_freq`. The function
+    handles special tokens that should be added to the beginning and end of the chunks.
+
+    Returns
+    -------
+    list of dictionaries that contain the chunked `input_ids`, `attention_mask`, `labels`
+
+    """
+
+    step = _get_chunking_step(max_length, overlap_freq)
+
+    def _chunked_seq(seq):
+        for i in range(len(seq) // step + 1):
+            if i * step < len(seq):
+                yield seq[i * step:i * step + max_length]
+
+    if labels is not None:
+        chunks = zip(*[_chunked_seq(seq) for seq in (input_ids, attention_mask, labels)])
+        chunks = (_prepare_for_model(*chunk, tokenizer=tokenizer) for chunk in chunks)
+        return [dict(input_ids=input_ids,
+                     attention_mask=attention_mask,
+                     labels=labels,
+                     offset=i * step,
+                     id=id,
+                     chunk_id=i)
+                for i, (input_ids, attention_mask, labels) in enumerate(chunks)]
+    else:
+        chunks = zip(*[_chunked_seq(seq) for seq in (input_ids, attention_mask)])
+        chunks = (_prepare_for_model(*chunk, labels=None, tokenizer=tokenizer) for chunk in chunks)
+        return [dict(input_ids=input_ids,
+                     attention_mask=attention_mask,
+                     offset=i * step,
+                     id=id,
+                     chunk_id=i)
+                for i, (input_ids, attention_mask, labels) in enumerate(chunks)]
+
+
+def _prepare_for_model(input_ids: List[int], attention_mask: List[int], labels: Optional[List[int]] = None, *,
+                       tokenizer: PreTrainedTokenizer, null_label_id: int = LABEL2ID['O']):
+    start_token_id = tokenizer.cls_token_id if tokenizer.cls_token_id is not None else tokenizer.bos_token_id
+    end_token_id = tokenizer.sep_token_id if tokenizer.sep_token_id is not None else tokenizer.eos_token_id
+
+    input_ids = [start_token_id] + list(input_ids) + [end_token_id]
+    attention_mask = [1] + list(attention_mask) + [1]
+    if labels is not None:
+        labels = [null_label_id] + list(labels) + [null_label_id]
```

Comment:
I think it's better to use -100 for the special tokens instead of `LABEL2ID['O']`